lib/puma/request.rb

```diff
@@ -69,127 +76,133 @@ def handle_request(client, lines, requests)
-          @log_writer.log "Detected force shutdown of a thread"
 
-          status, headers, res_body = lowlevel_error(e, env, 503)
-        rescue Exception => e
-          @log_writer.unknown_error e, client, "Rack app"
-
-          status, headers, res_body = lowlevel_error(e, env, 500)
-        end
-
-        res_info = {}
-        res_info[:content_length] = nil
-        res_info[:no_body] = head
-
-        res_info[:content_length] = if res_body.kind_of? Array and res_body.size == 1
-          res_body[0].bytesize
-        else
-          nil
+          return :async
         end
+      rescue ThreadPool::ForceShutdown => e
+        @log_writer.unknown_error e, client, "Rack app"
+        @log_writer.log "Detected force shutdown of a thread"
 
-        cork_socket io
+        status, headers, res_body = lowlevel_error(e, env, 503)
+      rescue Exception => e
+        @log_writer.unknown_error e, client, "Rack app"
 
-        str_headers(env, status, headers, res_info, lines, requests, client)
-
-        line_ending = LINE_END
-
-        content_length  = res_info[:content_length]
-        if res_body && !res_body.respond_to?(:each)
-          response_hijack = res_body
-        else
-          response_hijack = res_info[:response_hijack]
-        end
+        status, headers, res_body = lowlevel_error(e, env, 500)
+      end
+      write_response(status, headers, res_body, io_buffer, requests, client)
+    end
 
-        if res_info[:no_body]
-          if content_length and status != 204
-            lines.append CONTENT_LENGTH_S, content_length.to_s, line_ending
+    # Does the actual response writing for Request#handle_request and Server#client_error
+    #
+    # @param status [Integer] the status returned by the Rack application
+    # @param headers [Hash] the headers returned by the Rack application
+    # @param app_body [Array] the body returned by the Rack application or
+    #   a call to `lowlevel_error`
+    # @param io_buffer [Puma::IOBuffer] modified in place
+    # @param requests [Integer] number of inline requests handled
+    # @param client [Puma::Client]
+    # @return [Boolean,:async]
+    def write_response(status, headers, app_body, io_buffer, requests, client)
+      env = client.env
+      socket  = client.io
+
+      after_reply = env[RACK_AFTER_REPLY] || []
+
+      return false if closed_socket?(socket)
+
+      resp_info = str_headers(env, status, headers, app_body, io_buffer, requests, client)
+
+      # below converts app_body into body, dependent on app_body's characteristics, and
+      # resp_info[:content_length] will be set if it can be determined
+      if !resp_info[:content_length] && !resp_info[:transfer_encoding] && status != 204
+        if app_body.respond_to?(:to_ary)
+          length = 0
+          if array_body = app_body.to_ary
+            body = array_body.map { |part| length += part.bytesize; part }
+          elsif app_body.is_a?(::File) && app_body.respond_to?(:size)
```

Comment:
Perhaps we should check if app_body responds to `to_path` instead of `is_a?(::File)` to conform with the [Rack SPEC](https://github.com/rack/rack/blob/main/SPEC.rdoc#the-body-)? 

The SPEC mentions 'File-like' and `to_path` but doesn't mention the `File` class specifically.

Reply:
So I'm looking at [ActionDispatch::Response::FileBody](https://msp-greg.github.io/rails_master/ActionDispatch/Response/FileBody.html), and seeing that it is not a `File`, but responds to `to_path` and `each` is patched to return 16k chunks.  Since it's not a `File`, we can't get the size from it, so it has to be chunked?

Reply:
I added another commit to handle `to_path`.  Reason for the File object check is that Puma can use `copy_stream` for File objects, which would typically be faster than calling any kind of `each` iteration.

Reply:
Rack 3.0.0 released…